sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -0,0 +1,216 @@
+import org.apache.orc.storage.serde2.io.{DateWritable, HiveDecimalWritable}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecificInternalRow
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.datasources.orc.OrcUtils.withNullSafe
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+private[orc] class OrcDeserializer(
+    dataSchema: StructType,
+    requiredSchema: StructType,
+    missingColumnNames: Seq[String]) {
+
+  private[this] val mutableRow = new SpecificInternalRow(requiredSchema.map(_.dataType))
+
+  private[this] val length = requiredSchema.length
+
+  private[this] val unwrappers = requiredSchema.map(_.dataType).map(unwrapperFor).toArray
+
+  def deserialize(orcStruct: OrcStruct): InternalRow = {
+    var i = 0
+    val names = orcStruct.getSchema.getFieldNames
+    while (i < length) {
+      val name = requiredSchema(i).name
+      val writable = if (missingColumnNames.contains(name)) {
+        null
+      } else {
+        if (names.contains(name)) {
+          orcStruct.getFieldValue(name)
+        } else {
+          orcStruct.getFieldValue("_col" + dataSchema.fieldIndex(name))
+        }
+      }
+      if (writable == null) {
+        mutableRow.setNullAt(i)
+      } else {
+        unwrappers(i)(writable, mutableRow, i)
+      }
+      i += 1
+    }
+    mutableRow
+  }
+
+  private[this] def unwrapperFor(dataType: DataType): (Any, InternalRow, Int) => Unit =
+    dataType match {
+      case NullType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setNullAt(ordinal)
+
+      case BooleanType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setBoolean(ordinal, value.asInstanceOf[BooleanWritable].get)
+
+      case ByteType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setByte(ordinal, value.asInstanceOf[ByteWritable].get)
+
+      case ShortType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setShort(ordinal, value.asInstanceOf[ShortWritable].get)
+
+      case IntegerType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setInt(ordinal, value.asInstanceOf[IntWritable].get)
+
+      case LongType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setLong(ordinal, value.asInstanceOf[LongWritable].get)
+
+      case FloatType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setFloat(ordinal, value.asInstanceOf[FloatWritable].get)
+
+      case DoubleType =>
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row.setDouble(ordinal, value.asInstanceOf[DoubleWritable].get)
+
+      case _ =>
+        val unwrapper = getValueUnwrapper(dataType)
+        (value: Any, row: InternalRow, ordinal: Int) =>
+          row(ordinal) = unwrapper(value)
```

Comment:
how about
```
case s: StructType =>
  val types = s.map(_.dataType)
  val unwrappers = types.map(unwrapperFor)
  val mutableRow = new SpecificInternalRow(types)
  (value: Any, row: InternalRow, ordinal: Int) => {
    row(ordinal) = mutableRow
    val orcStruct = value.asInstanceOf[OrcStruct]
    var i = 0
    whie (i < unwrappers.length) {
      val fieldValue = orcStruct. getFieldValue(s(i).name)
      if (fieldValue == null) {
        mutableRow.setNullAt(i)
      } else {
        unwrapper(, mutableRow, i)
      }
      i += 1
     }
  }
```

Reply:
also add array and map type here, then we can remove `convertOrcStructToInternalRow` and `getValueUnwrapper`

Reply:
oops this approach doesn't work for array and map type. I think we should follow `ParquetRowConverter` and introduce a trait `OrcDataUpdater`, then we implement `StructDataUpdater`, `ArrayDataUpdater` and `MapDataUpdater`, and the returned function should be `(Any, OrcDataUpdater, Int) => Unit`

Reply:
@cloud-fan . The current way is an old ORC way. Do we need to introduce the Parquet way for some performance reason?

Reply:
And, for example, if the mappings look like the following, do we need to refactor some of the pattern between Parquet and ORC?
- `ArrayDataUpdater` <= `ParquetArrayConverter`
- `MapDataUpdater` <= `ParquetMapConverter`

Reply:
I feel Updater is more precise here, let's not follow the naming of parquet.

Reply:
it's not only for performance, but also remove duplicated code.

Reply:
I see. Thanks.